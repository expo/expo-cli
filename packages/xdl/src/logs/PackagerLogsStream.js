/* @flow */
import path from 'path';
import escapeStringRegexp from 'escape-string-regexp';
import chalk from 'chalk';
import getenv from 'getenv';
import { trim } from 'lodash';

import * as ProjectUtils from '../project/ProjectUtils';
import Logger from '../Logger';
import { isNode } from '../tools/EnvironmentHelper';

type ChunkT =
  | {
      tag: 'expo' | 'device',
      id: string,
      shouldHide: boolean,
      msg: any,
      level: number,
    }
  | {
      tag: 'metro',
      id: string,
      shouldHide: boolean,
      msg: ReportableEvent,
      level: number,
    };

type LogUpdater = (logState: Array<ChunkT>) => Array<ChunkT>;

type Error =
  | {
      originModulePath: string,
      message: string,
      errors: Array<Object>,
    }
  | {
      type: 'TransformError',
      snippet: string,
      lineNumber: number,
      column: number,
      filename: string,
      errors: Array<Object>,
    };

// Metro reporter types
// https://github.com/facebook/metro/blob/2a327fb19dd62169ab3ae9069011d8a599cfcf3e/packages/metro/src/lib/reporting.js
type GlobalCacheDisabledReason = 'too_many_errors' | 'too_many_misses';
type BundleDetails = {
  entryFile: string,
  platform: string,
  dev: boolean,
  minify: boolean,
  bundleType: string,
};
type ReportableEvent =
  | {
      port: ?number,
      // $FlowFixMe: $ReadOnlyArray not recognized
      projectRoots: $ReadOnlyArray<string>,
      type: 'initialize_started',
    }
  | {
      type: 'initialize_done',
    }
  | {
      type: 'initialize_failed',
      port: number,
      error: Error,
    }
  | {
      buildID: string,
      type: 'bundle_build_done',
    }
  | {
      buildID: string,
      type: 'bundle_build_failed',
    }
  | {
      buildID: string,
      bundleDetails: BundleDetails,
      type: 'bundle_build_started',
    }
  | {
      error: Error,
      type: 'bundling_error',
    }
  | {
      type: 'dep_graph_loading',
    }
  | {
      type: 'dep_graph_loaded',
    }
  | {
      buildID: string,
      type: 'bundle_transform_progressed',
      transformedFileCount: number,
      totalFileCount: number,
    }
  | {
      type: 'global_cache_error',
      error: Error,
    }
  | {
      type: 'global_cache_disabled',
      reason: GlobalCacheDisabledReason,
    }
  | {
      type: 'transform_cache_reset',
    }
  | {
      type: 'worker_stdout_chunk',
      chunk: string,
    }
  | {
      type: 'worker_stderr_chunk',
      chunk: string,
    }
  | {
      type: 'hmr_client_error',
      error: Error,
    };

export default class PackagerLogsStream {
  _projectRoot: string;
  _getCurrentOpenProjectId: () => any;
  _updateLogs: (updater: LogUpdater) => void;
  _logsToAdd: Array<ChunkT>;
  _bundleBuildChunkID: ?number;
  _onStartBuildBundle: ?Function;
  _onProgressBuildBundle: ?Function;
  _onFinishBuildBundle: ?Function;
  _onFailBuildBundle: ?Function;
  _getSnippetForError: ?(error: Error) => ?string;
  _bundleBuildStart: ?Date;

  _resetState = () => {
    this._logsToAdd = [];
  };

  constructor({
    projectRoot,
    getCurrentOpenProjectId,
    updateLogs,
    onStartBuildBundle,
    onProgressBuildBundle,
    onFinishBuildBundle,
    getSnippetForError,
  }: {
    projectRoot: string,
    getCurrentOpenProjectId?: () => any,
    updateLogs: (updater: LogUpdater) => void,
    onStartBuildBundle?: (chunk: any) => void,
    onProgressBuildBundle?: (progress: number, start: Date, chunk: any) => void,
    onFinishBuildBundle?: (error: ?string, start: ?Date, end: Date, chunk: any) => void,
    getSnippetForError?: (error: Error) => ?string,
  }) {
    this._resetState();
    this._projectRoot = projectRoot;
    this._getCurrentOpenProjectId = getCurrentOpenProjectId || (() => 1);
    this._updateLogs = updateLogs;

    // Optional properties in case the consumer wants to handle updates on
    // its own, eg: for a progress bar
    this._onStartBuildBundle = onStartBuildBundle;
    this._onProgressBuildBundle = onProgressBuildBundle;
    this._onFinishBuildBundle = onFinishBuildBundle;

    // Optional function for creating custom code frame snippet
    // (e.g. with terminal colors) from a syntax error.
    this._getSnippetForError = getSnippetForError;

    this._attachLoggerStream();
  }

  _attachLoggerStream() {
    let projectId = this._getCurrentOpenProjectId();

    ProjectUtils.attachLoggerStream(this._projectRoot, {
      stream: {
        write: chunk => {
          if (chunk.tag !== 'metro' && chunk.tag !== 'expo') {
            return;
          } else if (this._getCurrentOpenProjectId() !== projectId) {
            // TODO: We should be confident that we are properly unsubscribing
            // from the stream rather than doing a defensive check like this.
            return;
          }

          chunk = this._maybeParseMsgJSON(chunk);
          chunk = this._formatMsg(chunk);
          chunk = this._cleanUpNodeErrors(chunk);
          if (chunk.tag === 'metro') {
            this._handleMetroEvent(chunk);
          } else if (
            typeof chunk.msg === 'string' &&
            chunk.msg.match(/\w/) &&
            chunk.msg[0] !== '{'
          ) {
            this._enqueueAppendLogChunk(chunk);
          }
        },
      },
      type: 'raw',
    });
  }

  _handleMetroEvent(originalChunk: ChunkT) {
    const chunk = { ...originalChunk };
    let { msg } = chunk;

    if (typeof msg === 'string') {
      if (msg.includes('HTTP/1.1') && !getenv.boolish('EXPO_DEBUG', false)) {
        // Do nothing with this message - we want to filter out network requests logged by Metro.
      } else {
        // If Metro crashes for some reason, it may log an error message as a plain string to stderr.
        this._enqueueAppendLogChunk(chunk);
      }
      return;
    }
    if (/^bundle_/.test(msg.type)) {
      this._handleBundleTransformEvent(chunk);
      return;
    }
    switch (msg.type) {
      case 'initialize_started': // SDK >=23 (changed in Metro v0.17.0)
      case 'initialize_packager_started': // SDK <=22
        chunk._metroEventType = 'METRO_INITIALIZE_STARTED';
        chunk.msg = msg.port
          ? `Starting Metro Bundler on port ${msg.port}.`
          : 'Starting Metro Bundler.';
        break;
      case 'initialize_done': // SDK >=23 (changed in Metro v0.17.0)
      case 'initialize_packager_done': // SDK <=22
        chunk.msg = `Metro Bundler ready.`;
        break;
      case 'initialize_failed': // SDK >=23 (changed in Metro v0.17.0)
      case 'initialize_packager_failed': {
        // SDK <=22
        // $FlowFixMe
        let code = msg.error.code;
        chunk.msg =
          code === 'EADDRINUSE'
            ? `Metro Bundler can't listen on port ${msg.port}. The port is in use.`
            : `Metro Bundler failed to start. (code: ${code})`;
        break;
      }
      case 'bundling_error':
        chunk.msg =
          this._formatModuleResolutionError(msg.error) || this._formatBundlingError(msg.error);
        chunk.level = Logger.ERROR;
        break;
      case 'transform_cache_reset':
        chunk.msg =
          'Your JavaScript transform cache is empty, rebuilding (this may take a minute).';
        break;
      case 'hmr_client_error':
        chunk.msg = `A WebSocket client got a connection error. Please reload your device to get HMR working again.`;
        break;
      case 'global_cache_disabled':
        if (msg.reason === 'too_many_errors') {
          chunk.msg =
            'The global cache is now disabled because it has been failing too many times.';
        } else if (msg.reason === 'too_many_misses') {
          chunk.msg = `The global cache is now disabled because it has been missing too many consecutive keys.`;
        } else {
          chunk.msg = `The global cache is now disabled. Reason: ${msg.reason}`;
        }
        break;
      case 'worker_stdout_chunk':
        chunk.msg = this._formatWorkerChunk('stdout', msg.chunk);
        break;
      case 'worker_stderr_chunk':
        chunk.msg = this._formatWorkerChunk('stderr', msg.chunk);
        break;
      // Ignored events.
      case 'dep_graph_loading':
      case 'dep_graph_loaded':
      case 'global_cache_error':
        return;
      default:
        chunk.msg = `Unrecognized event: ${JSON.stringify(msg)}`;
        break;
    }
    this._enqueueAppendLogChunk(chunk);
  }

  // Any event related to bundle building is handled here
  _handleBundleTransformEvent = (chunk: any) => {
    let { msg } = chunk;

    if (msg.type === 'bundle_build_started') {
      chunk._metroEventType = 'BUILD_STARTED';
      this._handleNewBundleTransformStarted(chunk);
    } else if (msg.type === 'bundle_transform_progressed') {
      chunk._metroEventType = 'BUILD_PROGRESS';
      this._bundleBuildChunkID
        ? this._handleUpdateBundleTransformProgress(chunk)
        : this._handleNewBundleTransformStarted(chunk);
    } else if (msg.type === 'bundle_build_failed') {
      chunk._metroEventType = 'BUILD_FAILED';
      if (!this._bundleBuildChunkID) {
        // maybe?
      } else {
        this._handleUpdateBundleTransformProgress(chunk);
      }
    } else if (msg.type === 'bundle_build_done') {
      chunk._metroEventType = 'BUILD_DONE';
      if (!this._bundleBuildChunkID) {
        // maybe?
      } else {
        this._handleUpdateBundleTransformProgress(chunk);
      }
    } else {
      // Unrecognized bundle_* message!
      console.log('Unrecognized bundle_* message!', msg);
    }
  };

  _handleNewBundleTransformStarted(chunk: any) {
    if (this._bundleBuildChunkID) {
      return;
    }

    this._bundleBuildChunkID = chunk.id;
    this._bundleBuildStart = new Date();
    chunk.msg = 'Building JavaScript bundle';

    if (this._onStartBuildBundle) {
      this._onStartBuildBundle(chunk);
    } else {
      this._enqueueAppendLogChunk(chunk);
    }
  }

  _handleUpdateBundleTransformProgress(progressChunk: any) {
    let { msg } = progressChunk;
    let percentProgress;
    let bundleComplete = false;
    let bundleError = false;
    let bundleBuildEnd;

    if (msg.type === 'bundle_build_done') {
      percentProgress = 100;
      bundleComplete = true;
      bundleBuildEnd = new Date();
    } else if (msg.type === 'bundle_build_failed') {
      percentProgress = -1;
      bundleComplete = true;
      bundleError = new Error('Failed to build bundle');
      bundleBuildEnd = new Date();
    } else {
      percentProgress = Math.floor((msg.transformedFileCount / msg.totalFileCount) * 100);
    }

    if (this._bundleBuildChunkID) {
      progressChunk.id = this._bundleBuildChunkID;
    }
    if (bundleError) {
      progressChunk.msg = `Building JavaScript bundle: error`;
      progressChunk.level = Logger.ERROR;
      if (msg.error) {
        progressChunk.msg += '\n' + (msg.error.description || msg.error.message);
      }
    } else {
      if (bundleComplete) {
        let duration;
        if (this._bundleBuildStart) {
          duration = bundleBuildEnd - this._bundleBuildStart;
        }

        if (duration) {
          progressChunk.msg = `Building JavaScript bundle: finished in ${duration}ms.`;
        } else {
          progressChunk.msg = `Building JavaScript bundle: finished.`;
        }
      } else {
        progressChunk.msg = `Building JavaScript bundle: ${percentProgress}%`;
      }
    }

    if (this._onProgressBuildBundle) {
      this._onProgressBuildBundle(percentProgress, this._bundleBuildStart, progressChunk);

      if (bundleComplete) {
        this._onFinishBuildBundle &&
          this._onFinishBuildBundle(
            bundleError,
            this._bundleBuildStart,
            bundleBuildEnd,
            progressChunk
          );
        this._bundleBuildStart = null;
        this._bundleBuildChunkID = null;
      }
    } else {
      this._updateLogs(logs => {
        if (!logs || !logs.length) {
          return [];
        }

        logs.forEach(log => {
          if (log.id === this._bundleBuildChunkID) {
            log.msg = progressChunk.msg;
          }
        });

        if (bundleComplete) {
          this._bundleBuildChunkID = null;
        }

        return [...logs];
      });
    }
  }

  _formatModuleResolutionError(error: any) {
    let match = /^Unable to resolve module `(.+?)`/.exec(error.message);
    let { originModulePath } = error;
    if (!match || !originModulePath) {
      return null;
    }
    let moduleName = match[1];
    let relativePath = path.relative(this._projectRoot, originModulePath);

    const DOCS_PAGE_URL =
      'https://docs.expo.io/versions/latest/introduction/faq/#can-i-use-nodejs-packages-with-expo';

    if (NODE_STDLIB_MODULES.includes(moduleName)) {
      if (originModulePath.includes('node_modules')) {
        return `The package at "${relativePath}" attempted to import the Node standard library module "${moduleName}". It failed because React Native does not include the Node standard library. Read more at ${DOCS_PAGE_URL}`;
      } else {
        return `You attempted attempted to import the Node standard library module "${moduleName}" from "${relativePath}". It failed because React Native does not include the Node standard library. Read more at ${DOCS_PAGE_URL}`;
      }
    }
    return `Unable to resolve "${moduleName}" from "${relativePath}"`;
  }

  _formatBundlingError(error: any) {
    let message = error.message;
    if (!message && error.errors && error.errors.length) {
      message = error.errors[0].description;
    }

    // Before metro@0.29.0 the message may include the filename twice.
    // TODO(ville): Remove this once we drop support for react-native v0.54.4 or older.
    if (message && error.filename) {
      let escapedFilename = escapeStringRegexp(error.filename);
      message = message.replace(
        new RegExp(`Error in ${escapedFilename}: ${escapedFilename}:`),
        `Error in ${error.filename}:`
      );
    }

    message = chalk.red(message);

    let snippet = (this._getSnippetForError && this._getSnippetForError(error)) || error.snippet;
    if (snippet) {
      message += `\n${snippet}`;
    }
    return message;
  }

  _formatWorkerChunk(origin: 'stdout' | 'stderr', chunk: string) {
    const lines = chunk.split('\n');
    if (lines.length >= 1 && lines[lines.length - 1] === '') {
      lines.splice(lines.length - 1, 1);
    }
    return lines.map(line => `transform[${origin}]: ${line}`).join('\n');
  }

  _enqueueAppendLogChunk(chunk: ChunkT) {
    if (!chunk.shouldHide) {
      this._logsToAdd.push(chunk);
      this._enqueueFlushLogsToAdd();
    }
  }

  _enqueueFlushLogsToAdd = () => {
    let func = () => {
      this._updateLogs(logs => {
        if (this._logsToAdd.length === 0) {
          return logs;
        }

        let nextLogs = logs.concat(this._logsToAdd);
        this._logsToAdd = [];
        return nextLogs;
      });
    };

    if (isNode()) {
      func();
    } else {
      setImmediate(func);
    }
  };

  _maybeParseMsgJSON(chunk: ChunkT) {
    try {
      let parsedMsg = JSON.parse(chunk.msg);
      chunk.msg = parsedMsg;
    } catch (e) {
      // non-JSON message
    }

    return chunk;
  }

  _cleanUpNodeErrors = (chunk: ChunkT) => {
    if (typeof chunk.msg === 'object') {
      return chunk;
    }

    if (chunk.msg.match(/\(node:.\d*\)/)) {
      // Example: (node:13817) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): SyntaxError: SyntaxError /Users/brent/universe/apps/new-project-template/main.js: Unexpected token (10:6)
      // The first part of this is totally useless, so let's remove it.
      if (chunk.msg.match(/UnhandledPromiseRejectionWarning/)) {
        chunk.msg = chunk.msg.replace(/\(node:.*\(rejection .*\):/, '');
        if (chunk.msg.match(/SyntaxError: SyntaxError/)) {
          chunk.msg = chunk.msg.replace('SyntaxError: ', '');
        }
      } else if (chunk.msg.match(/DeprecationWarning/)) {
        chunk.msg = '';
      }
    }

    return chunk;
  };

  // This message is just noise
  // Fall back to the same formatting we did on SDK <= 15 before we had a custom
  // reporter class.
  _formatMsg(chunk) {
    if (typeof chunk.msg === 'object') {
      return chunk;
    }

    if (chunk.msg.match(/Looking for JS files in/)) {
      chunk.msg = '';
    } else if (chunk.msg.match(/^[\u001b]/)) {
      chunk.msg = '';
    }

    chunk.msg = chunk.msg.replace(/\[\w{2}m/g, '');
    chunk.msg = chunk.msg.replace(/\[2K/g, '');
    chunk.msg = trim(chunk.msg);
    return chunk;
  }
}

const NODE_STDLIB_MODULES = [
  'assert',
  'async_hooks',
  'buffer',
  'child_process',
  'cluster',
  'crypto',
  'dgram',
  'dns',
  'domain',
  'events',
  'fs',
  'http',
  'https',
  'net',
  'os',
  'path',
  'punycode',
  'querystring',
  'readline',
  'repl',
  'stream',
  'string_decoder',
  'tls',
  'tty',
  'url',
  'util',
  'v8',
  'vm',
  'zlib',
];
