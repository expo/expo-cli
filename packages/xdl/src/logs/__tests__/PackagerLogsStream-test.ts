import stripAnsi from 'strip-ansi';

import PackagerLogsStream from '../PackagerLogsStream';

jest.mock('../../internal', () => {
  const internal = jest.requireActual('../../internal');

  return {
    ...internal,
    ProjectUtils: {
      attachLoggerStream: jest.fn(),
    },
  };
});

function createBundlingErrorChunk(error): any {
  return {
    // @ts-ignore
    hostname: 'Evans-MacBook-Pro.local',
    pid: 86582,
    type: 'project',
    project: '/app',
    level: 30,
    tag: 'metro',
    msg: {
      error,
      type: 'bundling_error',
    },
    time: new Date('2021-10-18T20:06:02.209Z'),
    id: 'd1809510-304e-11ec-9b2e-5d348245f90d',
    v: 0,
  };
}

describe(PackagerLogsStream, () => {
  it(`formats a Metro internal error`, () => {
    const streamer = new PackagerLogsStream({
      getSnippetForError: jest.fn(() => {
        return '';
      }),
      updateLogs: jest.fn(),
    } as any);

    streamer._handleChunk(
      createBundlingErrorChunk({
        type: 'TransformError',
        lineNumber: 0,
        errors: [
          {
            description:
              "node_modules/react-native/Libraries/NewAppScreen/components/logo.png: Cannot read property 'length' of undefined",
            lineNumber: 0,
          },
        ],
        name: 'SyntaxError',
        message:
          "node_modules/react-native/Libraries/NewAppScreen/components/logo.png: Cannot read property 'length' of undefined",
        stack:
          "TypeError: Cannot read property 'length' of undefined\n" +
          '    at applyAssetDataPlugins (/app/node_modules/metro/src/Assets.js:182:25)\n' +
          '    at getAssetData (/app/node_modules/metro/src/Assets.js:178:16)\n' +
          '    at async Object.transform (/app/node_modules/metro-transform-worker/src/utils/assetTransformer.js:30:16)\n' +
          '    at async transformAsset (/app/node_modules/metro-transform-worker/src/index.js:371:18)\n' +
          '    at async Object.transform (/app/node_modules/metro-transform-worker/src/index.js:559:14)',
      })
    );

    expect(streamer._getSnippetForError).toHaveBeenCalledWith({
      errors: [
        {
          description:
            "node_modules/react-native/Libraries/NewAppScreen/components/logo.png: Cannot read property 'length' of undefined",
          lineNumber: 0,
        },
      ],
      lineNumber: 0,
      message:
        "node_modules/react-native/Libraries/NewAppScreen/components/logo.png: Cannot read property 'length' of undefined",
      name: 'SyntaxError',
      stack: expect.stringMatching(/at applyAssetDataPlugins/),
      //    `TypeError: Cannot read property 'length' of undefined
      //              at applyAssetDataPlugins (/app/node_modules/metro/src/Assets.js:182:25)
      //              at getAssetData (/app/node_modules/metro/src/Assets.js:178:16)
      //              at async Object.transform (/app/node_modules/metro-transform-worker/src/utils/assetTransformer.js:30:16)
      //              at async transformAsset (/app/node_modules/metro-transform-worker/src/index.js:371:18)
      //              at async Object.transform (/app/node_modules/metro-transform-worker/src/index.js:559:14)`,
      type: 'TransformError',
    });

    const msg = stripAnsi(streamer._logsToAdd[0].msg);

    expect(streamer._logsToAdd.length).toBe(1);
    // Since the message has formatting, use a snapshot to keep it in an expected format.
    expect(msg).toMatchSnapshot();
    // Title is expected.
    expect(msg).toMatch(
      /node_modules\/react-native\/Libraries\/NewAppScreen\/components\/logo.png: Cannot read property 'length' of undefined/
    );
    // Stack trace is added.
    expect(msg).toMatch(
      /at applyAssetDataPlugins \(\/app\/node_modules\/metro\/src\/Assets\.js:182:25\)/
    );
  });
  it(`formats an application code syntax error`, () => {
    const streamer = new PackagerLogsStream({
      updateLogs: jest.fn(),
    } as any);

    streamer._handleChunk(
      createBundlingErrorChunk({
        type: 'TransformError',
        lineNumber: 5,
        column: 0,
        filename: 'App.js',
        errors: [
          {
            description:
              'SyntaxError: /app/App.js: Unexpected token (5:0)\n' +
              '\n' +
              "\x1B[0m \x1B[90m 3 |\x1B[39m \x1B[36mimport\x1B[39m { \x1B[33mStyleSheet\x1B[39m\x1B[33m,\x1B[39m \x1B[33mText\x1B[39m\x1B[33m,\x1B[39m \x1B[33mView\x1B[39m } \x1B[36mfrom\x1B[39m \x1B[32m'react-native'\x1B[39m\x1B[33m;\x1B[39m\x1B[0m\n" +
              '\x1B[0m \x1B[90m 4 |\x1B[39m\x1B[0m\n' +
              '\x1B[0m\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 5 |\x1B[39m \x1B[33m>\x1B[39m\x1B[0m\n' +
              '\x1B[0m \x1B[90m   |\x1B[39m \x1B[31m\x1B[1m^\x1B[22m\x1B[39m\x1B[0m\n' +
              '\x1B[0m \x1B[90m 6 |\x1B[39m \x1B[36mexport\x1B[39m \x1B[36mdefault\x1B[39m \x1B[36mfunction\x1B[39m \x1B[33mApp\x1B[39m() {\x1B[0m\n' +
              '\x1B[0m \x1B[90m 7 |\x1B[39m   \x1B[36mreturn\x1B[39m (\x1B[0m\n' +
              '\x1B[0m \x1B[90m 8 |\x1B[39m     \x1B[33m<\x1B[39m\x1B[33mView\x1B[39m style\x1B[33m=\x1B[39m{styles\x1B[33m.\x1B[39mcontainer}\x1B[33m>\x1B[39m\x1B[0m',
            filename: 'App.js',
            lineNumber: 5,
          },
        ],
        name: 'SyntaxError',
        message:
          'SyntaxError: /app/App.js: Unexpected token (5:0)\n' +
          '\n' +
          "\x1B[0m \x1B[90m 3 |\x1B[39m \x1B[36mimport\x1B[39m { \x1B[33mStyleSheet\x1B[39m\x1B[33m,\x1B[39m \x1B[33mText\x1B[39m\x1B[33m,\x1B[39m \x1B[33mView\x1B[39m } \x1B[36mfrom\x1B[39m \x1B[32m'react-native'\x1B[39m\x1B[33m;\x1B[39m\x1B[0m\n" +
          '\x1B[0m \x1B[90m 4 |\x1B[39m\x1B[0m\n' +
          '\x1B[0m\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 5 |\x1B[39m \x1B[33m>\x1B[39m\x1B[0m\n' +
          '\x1B[0m \x1B[90m   |\x1B[39m \x1B[31m\x1B[1m^\x1B[22m\x1B[39m\x1B[0m\n' +
          '\x1B[0m \x1B[90m 6 |\x1B[39m \x1B[36mexport\x1B[39m \x1B[36mdefault\x1B[39m \x1B[36mfunction\x1B[39m \x1B[33mApp\x1B[39m() {\x1B[0m\n' +
          '\x1B[0m \x1B[90m 7 |\x1B[39m   \x1B[36mreturn\x1B[39m (\x1B[0m\n' +
          '\x1B[0m \x1B[90m 8 |\x1B[39m     \x1B[33m<\x1B[39m\x1B[33mView\x1B[39m style\x1B[33m=\x1B[39m{styles\x1B[33m.\x1B[39mcontainer}\x1B[33m>\x1B[39m\x1B[0m',
        stack:
          'SyntaxError: /app/App.js: Unexpected token (5:0)\n' +
          '\n' +
          "\x1B[0m \x1B[90m 3 |\x1B[39m \x1B[36mimport\x1B[39m { \x1B[33mStyleSheet\x1B[39m\x1B[33m,\x1B[39m \x1B[33mText\x1B[39m\x1B[33m,\x1B[39m \x1B[33mView\x1B[39m } \x1B[36mfrom\x1B[39m \x1B[32m'react-native'\x1B[39m\x1B[33m;\x1B[39m\x1B[0m\n" +
          '\x1B[0m \x1B[90m 4 |\x1B[39m\x1B[0m\n' +
          '\x1B[0m\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 5 |\x1B[39m \x1B[33m>\x1B[39m\x1B[0m\n' +
          '\x1B[0m \x1B[90m   |\x1B[39m \x1B[31m\x1B[1m^\x1B[22m\x1B[39m\x1B[0m\n' +
          '\x1B[0m \x1B[90m 6 |\x1B[39m \x1B[36mexport\x1B[39m \x1B[36mdefault\x1B[39m \x1B[36mfunction\x1B[39m \x1B[33mApp\x1B[39m() {\x1B[0m\n' +
          '\x1B[0m \x1B[90m 7 |\x1B[39m   \x1B[36mreturn\x1B[39m (\x1B[0m\n' +
          '\x1B[0m \x1B[90m 8 |\x1B[39m     \x1B[33m<\x1B[39m\x1B[33mView\x1B[39m style\x1B[33m=\x1B[39m{styles\x1B[33m.\x1B[39mcontainer}\x1B[33m>\x1B[39m\x1B[0m\n' +
          '    at Object._raise (/app/node_modules/@babel/parser/lib/index.js:541:17)\n' +
          '    at Object.raiseWithData (/app/node_modules/@babel/parser/lib/index.js:534:17)\n' +
          '    at Object.raise (/app/node_modules/@babel/parser/lib/index.js:495:17)\n' +
          '    at Object.unexpected (/app/node_modules/@babel/parser/lib/index.js:3550:16)\n' +
          '    at Object.parseExprAtom (/app/node_modules/@babel/parser/lib/index.js:11857:22)\n' +
          '    at Object.parseExprAtom (/app/node_modules/@babel/parser/lib/index.js:7456:20)\n' +
          '    at Object.parseExprSubscripts (/app/node_modules/@babel/parser/lib/index.js:11414:23)\n' +
          '    at Object.parseUpdate (/app/node_modules/@babel/parser/lib/index.js:11394:21)\n' +
          '    at Object.parseMaybeUnary (/app/node_modules/@babel/parser/lib/index.js:11369:23)\n' +
          '    at Object.parseMaybeUnaryOrPrivate (/app/node_modules/@babel/parser/lib/index.js:11183:61)',
      })
    );

    const msg = stripAnsi(streamer._logsToAdd[0].msg);

    expect(streamer._logsToAdd.length).toBe(1);
    // Since the message has formatting, use a snapshot to keep it in an expected format.
    expect(msg).toMatchSnapshot();
    // Title is expected.
    expect(msg).toMatch(/SyntaxError: \/app\/App.js: Unexpected token \(5:0\)/);
  });
});
