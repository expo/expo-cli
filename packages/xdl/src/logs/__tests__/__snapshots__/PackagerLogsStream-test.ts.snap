// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackagerLogsStream formats a Metro internal error 1`] = `
"node_modules/react-native/Libraries/NewAppScreen/components/logo.png: Cannot read property 'length' of undefined
TypeError: Cannot read property 'length' of undefined
    at applyAssetDataPlugins (/app/node_modules/metro/src/Assets.js:182:25)
    at getAssetData (/app/node_modules/metro/src/Assets.js:178:16)
    at async Object.transform (/app/node_modules/metro-transform-worker/src/utils/assetTransformer.js:30:16)
    at async transformAsset (/app/node_modules/metro-transform-worker/src/index.js:371:18)
    at async Object.transform (/app/node_modules/metro-transform-worker/src/index.js:559:14)"
`;

exports[`PackagerLogsStream formats an application code syntax error 1`] = `
"SyntaxError: /app/App.js: Unexpected token (5:0)

  3 | import { StyleSheet, Text, View } from 'react-native';
  4 |
> 5 | >
    | ^
  6 | export default function App() {
  7 |   return (
  8 |     <View style={styles.container}>"
`;
